restli-tools/src/main/java/com/linkedin/restli/tools/clientgen/FluentApiGenerator.java

```diff
@@ -117,69 +123,114 @@ public static void main(String[] args) throws Exception
-    final File targetDirectory = new File(targetDirectoryPath);
-
-
-    final RestSpecParser.ParseResult parseResult = parser.parseSources(sources);
-
-    final StringBuilder message = new StringBuilder();
-    for (CodeUtil.Pair<ResourceSchema, File> pair : parseResult.getSchemaAndFiles())
-    {
-      ResourceSchema resourceSchema = pair.first;
       // Skip unstructured data resources for client generation
       if (resourceSchema != null && ResourceEntityType.UNSTRUCTURED_DATA == resourceSchema.getEntityType())
       {
-        continue;
+        return null;
       }
 
       BaseResourceSpec spec = null;
       if (resourceSchema.hasCollection())
       {
         spec = new CollectionResourceSpec(resourceSchema,
             new TemplateSpecGenerator(schemaResolver),
-            pair.second.getPath(),
-            schemaResolver);
+            sourceIdlName,
+            schemaResolver, resourceSchema.getCollection().getIdentifier().getParams());
       }
       else if (resourceSchema.hasSimple())
       {
         spec = new SimpleResourceSpec(resourceSchema,
             new TemplateSpecGenerator(schemaResolver),
-            pair.second.getPath(),
+            sourceIdlName,
             schemaResolver);
       }
       else if (resourceSchema.hasAssociation())
       {
         spec = new AssociationResourceSpec(resourceSchema,
             new TemplateSpecGenerator(schemaResolver),
-            pair.second.getPath(),
+            sourceIdlName,
             schemaResolver);
       }
       else
       {
-        continue;
+        return null;
       }
-
       File packageDir = new File(targetDirectory, spec.getNamespace().toLowerCase().replace('.', File.separatorChar));
       packageDir.mkdirs();
       // Generate FluentClient impl
       File implFile = new File(packageDir, CodeUtil.capitalize(spec.getResource().getName()) + FLUENT_CLIENT_FILE_SUFFIX + ".java");
       // Generate Resource interface
-      // TODO: move to universal client generator where fits better
       File interfaceFile= new File(packageDir, CodeUtil.capitalize(spec.getResource().getName()) + ".java");
+
+      String resourcePath = getResourcePath(resourceSchema.getPath());
+      List<String> pathKeys = getPathKeys(resourcePath);
+      spec.setPathKeys(pathKeys);
+      spec.setAncestorResourceSpecs(new ArrayList<>(parentList));
+      List<BaseResourceSpec> childrenList = new LinkedList<>();
+      if (spec.getSubResources() != null)
+      {
+        for (ResourceSchema sub : spec.getSubResources())
+        {
+          parentList.add(spec);
+          BaseResourceSpec childSpec = generateFluentClientByResource(sub,
+              schemaResolver,
+              velocityEngine,
+              targetDirectory,
+              sourceIdlName,
+              parentList,
+              message);
+          if (childSpec != null)
+          {
+            childrenList.add(childSpec);
+          }
+          parentList.remove(parentList.size() - 1);
```

Comment:
I know var `parentList` means the current parent list without include the current resource spec, but keep adding and removing makes me wonder if it is worth to consider this way (without keep adding and removing)
```
if there is sub resource in this spec:
  create a new list of parent spec by parentList.append(spec)
  and then call the recursive function.
```